core/src/main/java/org/springframework/security/access/hierarchicalroles/RoleHierarchyImpl.java

```diff
@@ -99,6 +104,10 @@ public class RoleHierarchyImpl implements RoleHierarchy {
 	 */
 	private Map<String, Set<GrantedAuthority>> rolesReachableInOneOrMoreStepsMap = null;
 
+	public static Builder.RoleBranchBuilder role(String role) {
```

Comment:
I believe [this comment](https://github.com/spring-projects/spring-security/issues/13300#issuecomment-1606115443) raises something that this PR hasn't considered yet, which is role prefixes.

What if instead we went back to your earlier suggestion and did the following:

```java
public static Builder builder() {
    return new Builder("ROLE_");
}

public static Builder withRolePrefix(String rolePrefix) {
    return new Builder(rolePrefix);
}
```

The reason is that, in Spring Security, things like `hasRole("USER")` imply that `ROLE_` will be prefixed automatically, as opposed to `hasAuthority(`ROLE_USER`)` where the application must supply the prefix. `role` methods in this class should do the same thing.

I think this arrangement of methods is better aligned since they both return the same `Builder` class and we also don't introduce a `role` method that doesn't automatically prefix the underlying authority.

Reply:
Sure, makes sense, `implies` should also prefix ROLE to each one? Something like
```
role("FOO").implies("BAR")
```

resulting in: `ROLE_FOO > ROLE_BAR` ?

Reply:
Or whatever prefix is defined using `withRolePrefix`

I'm thinking maybe `builderWithRolePrefix` ? Or just overload `builder` with a parameter `rolePrefix`?

Just because `RoleHierarchyImpl.withRolePrefix(rolePrefix)` doesn't sound very intuitive to me. Wdyt?

Reply:
> `role("FOO").implies("BAR")`

Yes, that's the idea.

> Or whatever prefix is defined using `withRolePrefix`

Good question. The `with` prefix comes from a naming convention in Spring Security. `withXXX` means "give me a builder based on XXX" and `fromXXX` means "construct an instance using values XXX". An example is `NimbusJwtDecoder#withJwkSetUri`.

If what you are looking for is to align the method names, maybe we consider something like `withDefaults()` or `withDefaultRolePrefix()` instead of `builder()`. I kind of like the subtle reminder that `withDefaultRolePrefix()` will include the default role prefix for you.

Reply:
Liked the `withDefaultRolePrefix()`

Went ahead with this approach and implemented another one with no prefix, for cases where no standard prefix is used.

That led to 3 factory methods to build a hierarchy: `withDefaultRolePrefix`, `withRolePrefix` and `withNoRolePrefix`

Let me know if it looks good for you! Ty in advance